### DISCUSSION TITLE
How to chain multiple agents, dispatchers, and interceptors??

### DISCUSSION THREAD

**Discussion by SukkaW** [nodejs/undici#3721] [Q&A] [answered]
I noticed that there are many useful built-in agents/dispatchers/interceptors. I'd like to enable `EnvHttpProxyAgent` and `RetryAgent` both at once. I also want to enable automatic redirect and custom response error (and HTTP caching when it lands in the future) with interceptors, while enabling DNS lookup caching through `cacheable-lookup`. 

I am also a fan of `fetch`. How do I create a fetch function with all those features enabled?

**Comment by mcollina** [comment]
The Agent has a `compose` function for this purpose. Set the resulting value as the globalDispatcher.

**Reply by SukkaW** [reply]
> The Agent has a `compose` function for this purpose. Set the resulting value as the globalDispatcher.

Thanks for your reply!

I know `compose` exists. From what I see, `compose` is part of `Dispatcher`, while `Agent` extends `Dispatcher`. However, `Dispatcher#compose` accepts the `Dispatcher.DispatcherComposeInterceptor[]` type, but I can't see any `Agent` (`EnvHttpProxyAgent` for example) using that `Dispatcher.DispatcherComposeInterceptor` type.

Also there are `RetryAgent` (as an Agent) and retry interceptor, which one is recommended though?

**Reply by mcollina** [reply, marked as answer]
there is a retry interceptor you can use. We should be deprecate and remove the retryagent.

**Reply by SukkaW** [reply]
Thanks! I am able to create a global dispatcher:

```ts
const agent = new EnvHttpProxyAgent({
  allowH2: true,
  connect: {
    lookup(hostname, opt, cb) {
      return cacheableLookup.lookup(hostname, opt as CacheableLookupOptions, cb);
    }
  }
});

setGlobalDispatcher(agent.compose(
  interceptors.retry({
    maxRetries: 5,
    minTimeout: 10000
  }),
  interceptors.redirect({
    maxRedirections: 5
  }),
  interceptors.responseError({
    throwOnError: true
  })
));
```

According to my understanding, there should only be one `Agent`, no chained multiple `Agent`, is that right?

**Reply by mcollina** [reply]
yes, exactly.

**Reply by metcoder95** [reply]
You can also use the `dns` interceptor with a custom lookup; it requires some adjustments but overall through the interceptors you should be able to achieve the different strategies you highlighted here

**Reply by mbrevda** [reply]
> We should be deprecate and remove the retryagent

can you elaborate on that? Will the retry agent be removed from undici?

**Reply by metcoder95** [reply]
No. I was suggesting the usage of DNS interceptor with a custom lookup
